File: core/tauri/mobile/android/src/main/java/app/tauri/plugin/Invoke.kt
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package app.tauri.plugin

import app.tauri.Logger
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper

class Invoke(
  val id: Long,
  val command: String,
  val callback: Long,
  val error: Long,
  private val sendResponse: (callback: Long, data: String) -> Unit,
  private val argsJson: String,
  private val jsonMapper: ObjectMapper
) {
  fun<T> parseArgs(cls: Class<T>): T {
    return jsonMapper.readValue(argsJson, cls)
  }

  fun<T> parseArgs(ref: TypeReference<T>): T {
    return jsonMapper.readValue(argsJson, ref)
  }

  fun resolve(data: JSObject?) {
    sendResponse(callback, PluginResult(data).toString())
  }

  fun resolveObject(data: Any) {
    sendResponse(callback, jsonMapper.writeValueAsString(data))
  }

  fun resolve() {
    sendResponse(callback, "null")
  }

  fun reject(msg: String?, code: String?, ex: Exception?, data: JSObject?) {
    val errorResult = PluginResult()

    if (ex != null) {
      Logger.error(Logger.tags("Plugin"), msg!!, ex)
    }

    errorResult.put("message", msg)
    if (code != null) {
      errorResult.put("code", code)
    }
    if (data != null) {
      errorResult.put("data", data)
    }

    sendResponse(error, errorResult.toString())
  }

  fun reject(msg: String?, ex: Exception?, data: JSObject?) {
    reject(msg, null, ex, data)
  }

  fun reject(msg: String?, code: String?, data: JSObject?) {
    reject(msg, code, null, data)
  }

  fun reject(msg: String?, code: String?, ex: Exception?) {
    reject(msg, code, ex, null)
  }

  fun reject(msg: String?, data: JSObject?) {
    reject(msg, null, null, data)
  }

  fun reject(msg: String?, ex: Exception?) {
    reject(msg, null, ex, null)
  }

  fun reject(msg: String?, code: String?) {
    reject(msg, code, null, null)
  }

  fun reject(msg: String?) {
    reject(msg, null, null, null)
  }
}

File: crates/tauri/mobile/ios-api/Sources/Tauri/Invoke.swift
// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import Foundation
import UIKit

@objc public class Invoke: NSObject {
  public let command: String
  let callback: UInt64
  let error: UInt64
  let data: String
  let sendResponse: (UInt64, String?) -> Void
  let sendChannelData: (UInt64, String) -> Void

  public init(
    command: String, callback: UInt64, error: UInt64,
    sendResponse: @escaping (UInt64, String?) -> Void,
    sendChannelData: @escaping (UInt64, String) -> Void, data: String
  ) {
    self.command = command
    self.callback = callback
    self.error = error
    self.data = data
    self.sendResponse = sendResponse
    self.sendChannelData = sendChannelData
  }

  public func getRawArgs() -> String {
    return self.data
  }

  public func getArgs() throws -> JSObject {
    let jsonData = self.data.data(using: .utf8)!
    let data = try JSONSerialization.jsonObject(with: jsonData, options: [])
    return JSTypes.coerceDictionaryToJSObject(
      (data as! NSDictionary), formattingDatesAsStrings: true)!
  }

  public func parseArgs<T: Decodable>(_ type: T.Type) throws -> T {
    let jsonData = self.data.data(using: .utf8)!
    let decoder = JSONDecoder()
    decoder.userInfo[channelDataKey] = sendChannelData
    return try decoder.decode(type, from: jsonData)
  }

  func serialize(_ data: JsonValue) -> String {
    do {
      return try data.jsonRepresentation() ?? "\"Failed to serialize payload\""
    } catch {
      return "\"\(error)\""
    }
  }

  public func resolve() {
    sendResponse(callback, nil)
  }

  public func resolve(_ data: JsonObject) {
    resolve(.dictionary(data))
  }

  public func resolve(_ data: JsonValue) {
    sendResponse(callback, serialize(data))
  }

  public func resolve<T: Encodable>(_ data: T) {
    do {
      let json = try JSONEncoder().encode(data)
      sendResponse(callback, String(decoding: json, as: UTF8.self))
    } catch {
      sendResponse(self.error, "\"\(error)\"")
    }
  }

  public func reject(
    _ message: String, code: String? = nil, error: Error? = nil, data: JsonValue? = nil
  ) {
    let payload: NSMutableDictionary = [
      "message": message
    ]

    if let code = code {
      payload["code"] = code
    }

    if let error = error {
      payload["error"] = error
    }

    if let data = data {
      switch data {
      case .dictionary(let dict):
        for entry in dict {
          payload[entry.key] = entry.value
        }
      }
    }

    sendResponse(self.error, serialize(.dictionary(payload as! JsonObject)))
  }

  public func unimplemented() {
    unimplemented("not implemented")
  }

  public func unimplemented(_ message: String) {
    reject(message)
  }

  public func unavailable() {
    unavailable("not available")
  }

  public func unavailable(_ message: String) {
    reject(message)
  }
}
